packages/front-end/components/Settings/EditDataSource/DataSourceInlineEditIdentityJoins/AddEditIdentityJoinModal.tsx

```diff
@@ -0,0 +1,144 @@
+
+import { useForm } from "react-hook-form";
+import Modal from "../../../Modal";
+import MultiSelectField from "../../../Forms/MultiSelectField";
+import CodeTextArea from "../../../Forms/CodeTextArea";
+import { isDuplicateIdentityJoin } from "./utils";
+
+type AddEditIdentityJoinModalProps = {
+  identityJoin: IdentityJoinQuery | null;
+  dataSource: DataSourceInterfaceWithParams;
+  mode: "add" | "edit";
+  onSave: (identityJoin: IdentityJoinQuery) => void;
+  onCancel: () => void;
+};
+
+export const AddEditIdentityJoinModal: FC<AddEditIdentityJoinModalProps> = ({
+  identityJoin,
+  mode,
+  dataSource,
+  onCancel,
+  onSave,
+}) => {
+  const identityTypes = dataSource.settings.userIdTypes || [];
+  const existingIdentityJoins = dataSource.settings.queries.identityJoins || [];
+
+  const defaultQuery = useMemo(() => {
+    return (
+      "SELECT \n" +
+      identityTypes
+        .map(({ userIdType }) => `  ${userIdType} as ${userIdType}`)
+        .join(", \n") +
+      "\nFROM my_table"
+    );
+  }, [identityTypes]);
+
+  const form = useForm<IdentityJoinQuery>({
+    defaultValues: {
+      ids: identityJoin?.ids || [],
+      query: mode === "add" ? defaultQuery : identityJoin?.query || "",
+    },
+  });
+
+  const handleSubmit = form.handleSubmit(async (value) => {
+    onSave(value);
+
+    form.reset({
+      ids: [],
+      query: "",
+    });
+  });
+
+  // region Validation
+
+  const userEnteredIdentityJoinIds = form.watch("ids");
+  const userEnteredQuery = form.watch("query");
+  const userHasEnteredEnoughData =
+    userEnteredQuery && userEnteredIdentityJoinIds.length >= 2;
+
+  const isDuplicate = useMemo(() => {
+    return (
+      mode === "add" &&
+      isDuplicateIdentityJoin(userEnteredIdentityJoinIds, existingIdentityJoins)
+    );
+  }, [existingIdentityJoins, mode, userEnteredIdentityJoinIds]);
+
+  const saveEnabled = !isDuplicate && userHasEnteredEnoughData;
+
+  // endregion Validation
+
+  const modalTitle = useMemo(() => {
+    if (mode === "add") {
+      return "Add Identifier Join";
+    }
+
+    return `Edit Identifier Join: ${identityJoin.ids.join(" ↔ ")}`;
+  }, [mode, identityJoin]);
+
+  if (!identityJoin && mode === "edit") {
+    console.error(
+      "ImplementationError: identityJoin is required for Edit mode"
+    );
+    return null;
+  }
+
+  return (
+    <Modal
+      open={true}
+      submit={handleSubmit}
+      close={onCancel}
+      size="max"
+      header={modalTitle}
+      cta="Save"
+      ctaEnabled={saveEnabled}
+      autoFocusSelector="#id-modal-identify-joins-heading"
+    >
+      <h4 id="id-modal-identify-joins-heading">Identifier Join</h4>
+      <p>Queries that return a mapping between different identifier types</p>
+
+      <div className="row">
+        <div className="col-xs-12 col-md-6">
+          <MultiSelectField
+            label="Identifier Types"
+            value={form.watch("ids")}
+            disabled={mode === "edit"}
```

Comment:
If disabling this makes the code a lot simpler, I'm fine with it as is.  Otherwise, it might be nice to allow editing this.  Nothing references identity join queries, so no danger of breaking things.  One specific use case I can think of is when a user adds a  new userIdType, they might want to just add it to an existing join query as well.